### PR TITLE
[webpack] Add RuntimeValue for DefinePlugin

### PR DESCRIPTION
**Pull request by danielthank**
Add type for https://github.com/webpack/webpack/issues/9316
This feature was first introduced in https://github.com/webpack/webpack/pull/6793

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/webpack/webpack/blob/627510d7e72582463d76608fe3cda5a70ea17a12/lib/DefinePlugin.js#L86>>

### CODE REVIEW

File: types/webpack/index.d.ts [hunk 1]
```diff
@@ -1623,6 +1624,7 @@ declare namespace webpack {
 
     class DefinePlugin extends Plugin {
         constructor(definitions: { [key: string]: any });
+        static runtimeValue(fn: ({ module }: { module: any }) => any, fileDependencies?: any[]): any;
```
Comment (johnnyreilly) [inline types/webpack/index.d.ts]: Can we be more specific than `any` here?

**Review by johnnyreilly — CHANGES_REQUESTED** [review]

**Comment by danielthank** [comment]
@johnnyreilly I changed the types and tests to be more specific. Please review again!

File: types/webpack/index.d.ts [hunk 2]
```diff
@@ -1623,8 +1623,22 @@ declare namespace webpack {
...
+        ): DefinePlugin.RuntimeValue;
+    }
+
+    namespace DefinePlugin {
+        class RuntimeValue {
+            constructor(
+                fn: ({ module }: { module: compilation.Module }) => CodeValuePrimitive,
+                fileDependencies?: string[]
+            );
+            exec(parser: any): CodeValuePrimitive;
```
Comment (johnnyreilly) [inline types/webpack/index.d.ts]: Nice work! Can `parser` be absolutely anything?

**Review by johnnyreilly — CHANGES_REQUESTED** [review]

**Review by johnnyreilly — APPROVED** [review]
Nice!

File: types/webpack/index.d.ts [hunk 3]
```diff
@@ -1622,7 +1665,22 @@ declare namespace webpack {
...
+            fn: ({ module }: { module: compilation.Module }) => DefinePlugin.CodeValuePrimitive,
+            fileDependencies?: string[]
+        ): DefinePlugin.RuntimeValue;
+    }
+
+    namespace DefinePlugin {
+        class RuntimeValue {
+            constructor(
+                fn: ({ module }: { module: compilation.Module }) => CodeValuePrimitive,
+                fileDependencies?: string[]
```
Comment (fregante) [inline types/webpack/index.d.ts]: This should be

```ts
fileDependencies?: true | string[]
```

https://github.com/webpack/webpack/blob/89c752411e9bd803401acf1b8402559dae4c6eea/lib/DefinePlugin.js#L30

File: types/webpack/index.d.ts [hunk 4]
```diff
@@ -1622,7 +1665,22 @@ declare namespace webpack {
...
+        static runtimeValue(
+            fn: ({ module }: { module: compilation.Module }) => DefinePlugin.CodeValuePrimitive,
+            fileDependencies?: string[]
+        ): DefinePlugin.RuntimeValue;
+    }
+
+    namespace DefinePlugin {
+        class RuntimeValue {
+            constructor(
+                fn: ({ module }: { module: compilation.Module }) => CodeValuePrimitive,
```
Comment (fregante) [inline types/webpack/index.d.ts]: `module` should have a `resource` property but doesn't:


```
webpack.config.ts:115:48 - error TS2551: Property 'resource' does not exist on type 'Module'. Did you mean '_source'?

115                     return JSON.stringify(path.basename(module.resource, '.tsx'));
                                                                   ~~~~~~~~

  node_modules/@types/webpack/index.d.ts:800:13
    800             _source: any;
                    ~~~~~~~
    '_source' is declared here.
```

This code has been working for a while

**Comment by fregante** [comment]
Thank you for this PR @johnnyreilly!

**Pull request merged**